### PR TITLE
Add log handler for custom handling

### PR DESCRIPTION
**Pull request by a-zakir**
This allows a user of Xpress/Coin to manage logs from solvers via a simple interface 
#TODO
- [x] Xpress Interface
- [x] COIN Interface
- [ ] Sirius Interface (postponed)
- [ ] maybe update Sirius version (postponed)

### CODE REVIEW

**Comment by flomnes** [comment]
Possible improvements
- Give `EnableOutput` an argument of type `std::ofstream` with default value `std::cout`
- Let the caller (the application) determine where to write, i.e remove `std::thread::get_id`, etc.
- Implement this logic for other solvers (COIN, etc.)

File: ortools/linear_solver/xpress_interface.cc [hunk 1]
```diff
@@ -2031,8 +2037,13 @@ void XPRS_CC optimizermsg(XPRSprob prob, void* data, const char* sMsg, int nLen,
...
       case 2: /* dialogue */
       case 1: /* information */
-        printf("%*s\n", nLen, sMsg);
-        break;
+      {
+        if (auto log_handler = xprs->solver_log_handler(); log_handler) {
+          log_handler->message(sMsg);
+        }
+      }
+      // printf("%*s\n", nLen, sMsg);
```
Comment (flomnes) [inline ortools/linear_solver/xpress_interface.cc:2045]: Use this as the default behavior ? Allow the user to print logs in `std::cout` easily
```cpp
if (/* ... */; log_handler) {
 // ...
} else {
    printf(/* ... */);
}
```

File: ortools/linear_solver/xpress_interface.cc [hunk 2]
```diff
@@ -1711,8 +1711,14 @@ MPSolver::ResultStatus XpressInterface::Solve(MPSolverParameters const& param) {
...
 
-  // Set log level.
-  XPRSsetintcontrol(mLp, XPRS_OUTPUTLOG, quiet() ? 0 : 1);
+  // Set log.
+  if (!quiet()){
+    XPRSsetintcontrol(mLp, XPRS_OUTPUTLOG, 1);
+}else {
+
+  XPRSsetintcontrol(mLp, XPRS_OUTPUTLOG, 0);
+}
```
[the remark this answers is not in the document]
Comment (a-zakir) [reply ortools/linear_solver/xpress_interface.cc:1720]: there were several lines in the previous version, that's why i created the if/else statement. [see](https://github.com/rte-france/or-tools/pull/112/commits/94f7dda9addf185579bf5b84257639efa8cfef32#r1396876817)

[the remark this answers is not in the document]
Comment (a-zakir) [reply ortools/linear_solver/xpress_interface.cc:1720]: i will clean

**Review by flomnes — COMMENTED** [review]
Maybe we need another `flush` method for interface `LogHandlerInterface`. The XPRESS doc indicates that `nMsgLvl` <0 indicates that it is necessary to synchronize logs, probably before a crash ?

**Comment by a-zakir** [comment]
> Maybe we need another `flush` method for interface `LogHandlerInterface`. The XPRESS doc indicates that `nMsgLvl` <0 indicates that it is necessary to synchronize logs, probably before a crash ?

 that's a good point! We've outsourced message processing to the handler, so it's up to the caller to manage it.
 changes are coming ....

**Comment by a-zakir** [comment]
> > Maybe we need another `flush` method for interface `LogHandlerInterface`. The XPRESS doc indicates that `nMsgLvl` <0 indicates that it is necessary to synchronize logs, probably before a crash ?
> 
> that's a good point! We've outsourced message processing to the handler, so it's up to the caller to manage it. changes are coming ....
i'm taking care of the default handler...

**Pull request closed without merging**